File: ssl-examiner/scan.py
import sys
import ssl
import socket
import json
from datetime import datetime


if len(sys.argv) < 2:
    print("\033[0;31m" + 'Missing argument <url>' + "\033[0m")
    sys.exit()

target = sys.argv[1]
verify_peer = '--verify-peer' in sys.argv
verbose_output = '--vvv' in sys.argv
red = "\033[0;31m"
green = "\033[0;32m"
reset = "\033[0m"

def curl_req(url, verify_peer):
    hostname = url.split("://")[-1].split(":")[0]
    context = ssl.create_default_context()
    ssl_socket = context.wrap_socket(socket.socket(socket.AF_INET), server_hostname=hostname)
    try:
        ssl_socket.connect((hostname, 443))
        ssl_cert = ssl_socket.getpeercert()
        # Extracting commonName if present
        common_name = ""
        subject = ssl_cert.get("subject", [])
        if isinstance(subject, list):
            for name, value in subject:
                if name == "commonName":
                    common_name = value
                    break

        issuer = ssl_cert.get("issuer", ())
        primary_issuer = ""
        if isinstance(issuer, tuple):
            for name, value in issuer:
                if name == "C":
                    primary_issuer = value
                    break

        ssl_cert_details = {
            "commonName": common_name,
            "validFrom_time_t": datetime.strptime(ssl_cert["notBefore"], "%b %d %H:%M:%S %Y %Z").timestamp(),
            "validTo_time_t": datetime.strptime(ssl_cert["notAfter"], "%b %d %H:%M:%S %Y %Z").timestamp(),
            "primary_issuer": primary_issuer,
        }
        return ssl_cert_details
    finally:
        ssl_socket.close()


ssl_certificate_details = curl_req(target, verify_peer)
valid = ssl_certificate_details['validFrom_time_t']
expires = ssl_certificate_details['validTo_time_t']
issuer = ssl_certificate_details['issuer']  # C, O, CN

if verbose_output:
    print(json.dumps(ssl_certificate_details, indent=4))

name = ssl_certificate_details['commonName']

json_data = {
    'name': name,
    'expires': datetime.fromtimestamp(expires).strftime('%H:%M:%S %d-%m-%Y'),
    'issued': datetime.fromtimestamp(valid).strftime('%H:%M:%S %d-%m-%Y'),
    'primary_issuer': issuer['C'],
    'secondary_issuer': issuer['O'],
    'ca_issuer': issuer['CN']
}

with open('report-' + name + '.txt', 'w') as report_file:
    json.dump(json_data, report_file)

print(green + 'Successfully saved to: report-' + name + '.txt' + reset)

if datetime.now().timestamp() > expires:
    print(red + 'SSL cert has expired' + reset)
else:
    print(green + 'SSL is valid' + reset)

print()

File: ssl/scan.php
<?php
$text_art = '
                ▒▒▒▒▒▒▒█▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀█
                ▒▒▒▒▒▒▒█░▒▒▒▒▒▒▒▓▒▒▓▒▒▒▒▒▒▒░█
                ▒▒▒▒▒▒▒█░▒▒▓▒▒▒▒▒▒▒▒▒▄▄▒▓▒▒░█░▄▄
                ▒▒▄▀▀▄▄█░▒▒▒▒▒▒▓▒▒▒▒█░░▀▄▄▄▄▄▀░░█
                ▒▒█░░░░█░▒▒▒▒▒▒▒▒▒▒▒█░░░░░░░░░░░█  <HOLD ON! HERE WE GO!
                ▒▒▒▀▀▄▄█░▒▒▒▒▓▒▒▒▓▒█░░░█▒░░░░█▒░░█
                ▒▒▒▒▒▒▒█░▒▓▒▒▒▒▓▒▒▒█░░░░░░░▀░░░░░█
                ▒▒▒▒▒▄▄█░▒▒▒▓▒▒▒▒▒▒▒█░░█▄▄█▄▄█░░█
                ▒▒▒▒█░░░█▄▄▄▄▄▄▄▄▄▄█░█▄▄▄▄▄▄▄▄▄█
                ▒▒▒▒█▄▄█░░█▄▄█░░░░░░█▄▄█░░█▄▄█
';

echo  $text_art . PHP_EOL;
$target = $argv[1];
$verifyPeer = in_array('--verify-peer', $argv);
$verboseOutput = in_array('--vvv', $argv);
$red = "\033[0;31m";
$green = "\033[0;32m";
$reset = "\033[0m"; // Reset to default color

function curlReq(string $url, bool $verifyPeer): array
{
    if (str_contains($url, '://')) {
        $url = str_replace(['https://', 'http://', '://'], '', $url);
    }

    $context = stream_context_create([
        "ssl" => [
            "capture_peer_cert" => true,
            "verify_peer" => $verifyPeer
        ]
    ]);

    $stream = stream_socket_client("ssl://$url:443",
        $errno,
        $errstr,
        30,
        STREAM_CLIENT_CONNECT,
        $context
    );

    $sslContext = stream_context_get_params($stream);
    $sslCertificate = $sslContext["options"]["ssl"]["peer_certificate"];
    $sslCertDetails = openssl_x509_parse($sslCertificate);

    fclose($stream);

    return $sslCertDetails;
}

if (!isset($argv[1])) {
    exit($red . 'Missing argument <url>' . $reset . PHP_EOL);
}

$sslCertificateDetails = curlReq($argv[1], $verifyPeer);
$valid = $sslCertificateDetails['validFrom_time_t'];
$expires = $sslCertificateDetails['validTo_time_t'];
$issuer = $sslCertificateDetails['issuer']; // C, O, CN
if ($verboseOutput) {
    print_r($sslCertificateDetails);
}

$name = explode('=', $sslCertificateDetails['name']);

$json = [
    'name' => $name[1] ?? $sslCertificateDetails['name'],
    'expires' => date('h:i:s d-m-Y', $expires),
    'issued' => date('h:i:s d-m-Y', $valid),
    'primary_issuer' => $issuer['C'],
    'Secondary_issuer' => $issuer['O'],
    'CA_issuer' => $issuer['CN']
];


if (file_put_contents('report-' . $name[1] ?? $sslCertificateDetails['name'] . '.txt', json_encode($json))) {
    echo $green . PHP_EOL . 'Successfully saved to: ' . $name[1] ?? $sslCertificateDetails['name'] . '.txt' . $reset;
}

if (time() > $expires) {
    echo $red . PHP_EOL . 'SSL cert has expired' . $reset;
} else {
    echo $green . PHP_EOL .'SSL is valid' . $reset;
}

echo PHP_EOL;